File: app/src/main/java/com/mogralabs/mogra/audio/Melody.kt
package com.mogralabs.mogra.audio

import kotlin.math.exp
import kotlin.math.floor
import kotlin.math.ln
import kotlin.math.max
import kotlin.math.pow
import kotlin.math.sqrt

/**
 * The pitch-histogram branch — a Kotlin transcription of `model/portable/melody.py`.
 *
 * CREPE gives a frame-level f0 track; every voiced frame is expressed in cents above Sa,
 * folded into one octave and dropped into 120 bins, blurred slightly so two performances
 * tuned a few cents apart still overlap, and raised to the power 0.5 so one long held nyas
 * note cannot swamp every other swar.
 *
 * Two inherited quirks that are invisible if you get them wrong:
 *
 * * the exported net already ends in a sigmoid, and the decode applies a **second** one
 *   before weighting — the histograms the linear model was fitted on came through both;
 * * the 360 bin centres are **dithered**, by a fixed draw that lives in `bin_cents.bin`.
 *   It is a constant, not something to regenerate.
 *
 * The hop is 40 ms rather than the 10 ms the track was originally computed at. Frames at
 * hop 2h are exactly the even-indexed frames at hop h, so a coarser hop is a uniform
 * subsample of the same track. Measured over the 150-clip test set the fused scores are
 * 10 ms → 0.480/0.820, 20 ms → 0.480/0.820, 40 ms → 0.487/0.820, 80 ms → 0.467/0.807. At
 * 40 ms it is a quarter of the work for no loss; 80 ms is where the histogram stops
 * separating a swar from the meend leading into it.
 */
object Melody {

    const val SR = 16000
    const val HOP = 640                 // 40 ms
    const val HOP_TRAINED = 160         // 10 ms, what the histograms were computed at
    const val WINDOW = 1024
    const val PITCH_BINS = 360
    const val CONFIDENCE = 0.4f
    const val N_BINS = 120
    private const val SMOOTH = 1.0
    private const val POWER = 0.5
    private const val CENTS_PER_BIN = 20.0
    private const val CENTS_0 = 1997.3794084376191
    private const val FMIN = 50.0
    private const val FMAX = 2000.0

    /** Frames of [WINDOW] samples every [hop], mean-centred and scaled by their own s.d. */
    fun frames(y16000: FloatArray, hop: Int = HOP): Array<FloatArray> {
        val total = 1 + y16000.size / hop
        val padded = FloatArray(y16000.size + WINDOW)
        System.arraycopy(y16000, 0, padded, WINDOW / 2, y16000.size)
        return Array(total) { i ->
            val f = FloatArray(WINDOW)
            System.arraycopy(padded, i * hop, f, 0, WINDOW)
            var mean = 0.0
            for (v in f) mean += v
            mean /= WINDOW
            var ss = 0.0
            for (j in 0 until WINDOW) {
                f[j] = (f[j] - mean).toFloat()
                ss += f[j].toDouble() * f[j]
            }
            val sd = max(1e-10, sqrt(ss / (WINDOW - 1)))     // unbiased, as torch.std is
            for (j in 0 until WINDOW) f[j] = (f[j] / sd).toFloat()
            f
        }
    }

    private fun frequencyToBin(hz: Double, ceil: Boolean): Int {
        val cents = 1200.0 * (ln(hz / 10.0) / ln(2.0))
        val b = (cents - CENTS_0) / CENTS_PER_BIN
        return if (ceil) Math.ceil(b).toInt() else floor(b).toInt()
    }

    /**
     * Network output → (f0 in Hz, voiced mask).
     *
     * [probs] is `nFrames × 360`, already sigmoid, exactly as the exported module returns.
     */
    fun decode(probs: Array<FloatArray>, binCents: FloatArray): Pair<FloatArray, BooleanArray> {
        val lo = frequencyToBin(FMIN, ceil = false)
        val hi = frequencyToBin(FMAX, ceil = true)
        val f0 = FloatArray(probs.size)
        val voiced = BooleanArray(probs.size)

        for (t in probs.indices) {
            val p = probs[t]
            var best = lo
            for (b in lo until hi) if (p[b] > p[best]) best = b
            voiced[t] = p[best] >= CONFIDENCE

            val from = max(0, best - 4)
            val to = minOf(PITCH_BINS, best + 5)
            var num = 0.0
            var den = 0.0
            for (b in from until to) {
                if (b < lo || b >= hi) continue          // masked out before the decode
                val s = 1.0 / (1.0 + exp(-p[b].toDouble()))   // the second sigmoid
                num += s * binCents[b]
                den += s
            }
            val cents = if (den > 0) num / den else binCents[best].toDouble()
            f0[t] = (10.0 * 2.0.pow(cents / 1200.0)).toFloat()
        }
        return f0 to voiced
    }

    /** Voiced frames → a 120-bin octave-folded pitch histogram summing to 1. */
    fun histogram(f0: FloatArray, voiced: BooleanArray, tonicHz: Double): FloatArray {
        val h = DoubleArray(N_BINS)
        var n = 0
        for (i in f0.indices) {
            if (!voiced[i]) continue
            val f = f0[i].toDouble()
            if (f <= 0 || !f.isFinite()) continue
            val cents = 1200.0 * (ln(f / tonicHz) / ln(2.0))
            if (!cents.isFinite()) continue
            var folded = cents % 1200.0
            if (folded < 0) folded += 1200.0
            var idx = floor(folded * (N_BINS / 1200.0)).toInt() % N_BINS
            if (idx < 0) idx += N_BINS
            h[idx] += 1.0
            n++
        }
        if (n < 5) return FloatArray(N_BINS)

        // circular Gaussian blur, done directly -- 120 bins does not need an FFT
        val kern = DoubleArray(N_BINS) { val d = minOf(it, N_BINS - it); exp(-0.5 * (d / SMOOTH).pow(2)) }
        var ks = 0.0
        for (v in kern) ks += v
        val blurred = DoubleArray(N_BINS)
        for (i in 0 until N_BINS) {
            var acc = 0.0
            for (j in 0 until N_BINS) acc += h[j] * kern[((i - j) % N_BINS + N_BINS) % N_BINS]
            blurred[i] = max(0.0, acc / ks)
        }
        var total = 0.0
        for (i in 0 until N_BINS) { blurred[i] = blurred[i].pow(POWER); total += blurred[i] }
        return FloatArray(N_BINS) { if (total > 0) (blurred[it] / total).toFloat() else 0f }
    }

    /**
     * Standardise, then multinomial logistic regression — four arrays, not a pickled
     * scikit-learn object, so nothing here is tied to a library version.
     */
    class LinearModel(
        val mean: FloatArray, val scale: FloatArray,
        val coef: Array<FloatArray>, val intercept: FloatArray,
    ) {
        fun scores(hist: FloatArray): DoubleArray = DoubleArray(intercept.size) { r ->
            var acc = intercept[r].toDouble()
            val row = coef[r]
            for (b in hist.indices) acc += row[b] * (hist[b] - mean[b]) / scale[b]
            acc
        }
    }
}

File: model/portable/melody.py
"""The melody branch without torchcrepe.

torchcrepe is a thin wrapper: a conv net, and around it some framing, a double sigmoid, a
weighted-argmax decode and a dither. The net exports to ONNX or TFLite untouched. What has
to be rewritten for a phone is everything around it, and everything around it is here, in
numpy.

Three details are worth stating because getting any of them wrong is silent:

* **The double sigmoid.** `Crepe.forward` already ends in a sigmoid, and
  `torchcrepe.decode.weighted_argmax` applies another one to the same numbers before
  weighting. That is a quirk of the library, not of CREPE, but the histogram the linear
  model was fitted on came through both, so both are here.

* **The dither is drawn once per process, not once per clip.** It perturbs the 360 bin
  *centres*, and torchcrepe caches the perturbed vector on the function object. Every clip
  in a run therefore shares one draw. Seeded, so it is the same draw every run.

* **Periodicity is read before the decode masking**, from the single sigmoid, at the
  argmax bin.

**On the hop length.** The reference tracks pitch every 10 ms, which is 2001 frames for a
20 s window and 14.7 s of single-threaded CREPE -- the whole pipeline's cost, near enough.
Frames at hop 2h are exactly the even-indexed frames at hop h, so a coarser hop is a
uniform subsample of the same track, and the histogram downstream is a normalised
distribution that does not care how many frames it was built from. Measured on the 150
test clips, fused as usual:

    10 ms   top1 0.4800   top5 0.8200      (the trained setting)
    20 ms   top1 0.4800   top5 0.8200
    40 ms   top1 0.4867   top5 0.8200      <- default here, a quarter of the cost
    80 ms   top1 0.4667   top5 0.8067      <- too coarse, starts losing swars

40 ms is not better than 10 ms in any real sense -- one clip out of 150 is noise -- it is
free. 80 ms is where the histogram stops resolving the distinction between a swar and the
meend leading into it. If anything downstream ever wants the dense track, pass
`hop=HOP_TRAINED`.
"""

import functools
from pathlib import Path

import numpy as np
import torch
import torch.nn.functional as F

SR = 16000
HOP = 640                 # 40 ms -- see the note on hop length below
HOP_TRAINED = 160         # 10 ms, what the histograms were originally computed at
WINDOW = 1024
PITCH_BINS = 360
CENTS_PER_BIN = 20
CENTS_0 = 1997.3794084376191
CONFIDENCE = 0.4
FMIN, FMAX = 50.0, 2000.0
N_BINS = 120
SMOOTH = 1.0
POWER = 0.5
WEIGHTS = Path(__file__).resolve().parent / "crepe_tiny.pth"


class Crepe(torch.nn.Module):
    """CREPE-tiny, structurally identical to torchcrepe's so its checkpoint loads."""

    def __init__(self):
        super().__init__()
        in_channels = [1, 128, 16, 16, 16, 32]
        out_channels = [128, 16, 16, 16, 32, 64]
        self.in_features = 256
        kernel_sizes = [(512, 1)] + 5 * [(64, 1)]
        strides = [(4, 1)] + 5 * [(1, 1)]
        bn = functools.partial(torch.nn.BatchNorm2d, eps=0.0010000000474974513, momentum=0.0)
        for i in range(6):
            setattr(self, f"conv{i + 1}", torch.nn.Conv2d(
                in_channels[i], out_channels[i], kernel_sizes[i], strides[i]))
            setattr(self, f"conv{i + 1}_BN", bn(num_features=out_channels[i]))
        self.classifier = torch.nn.Linear(self.in_features, PITCH_BINS)

    def _layer(self, x, conv, batch_norm, padding=(0, 0, 31, 32)):
        x = F.relu(conv(F.pad(x, padding)))
        return F.max_pool2d(batch_norm(x), (2, 1), (2, 1))

    def forward(self, x):
        x = x[:, None, :, None]
        x = self._layer(x, self.conv1, self.conv1_BN, (0, 0, 254, 254))
        for i in range(2, 6):
            x = self._layer(x, getattr(self, f"conv{i}"), getattr(self, f"conv{i}_BN"))
        x = self._layer(x, self.conv6, self.conv6_BN)
        x = x.permute(0, 2, 1, 3).reshape(-1, self.in_features)
        return torch.sigmoid(self.classifier(x))


@functools.lru_cache(maxsize=1)
def _model(device="cpu"):
    net = Crepe()
    net.load_state_dict(torch.load(WEIGHTS, map_location=device, weights_only=True))
    return net.to(device).eval()


@functools.lru_cache(maxsize=4)
def _bin_cents(dither_seed=0):
    """The 360 bin centres in cents, dithered once -- see the module docstring.

    scipy.stats.triang.rvs(c=0.5, loc=-20, scale=40) is numpy's `triangular(0, 0.5, 1)`
    shifted and scaled, so this reproduces torchcrepe's draw without scipy.
    """
    state = np.random.get_state()
    try:
        np.random.seed(dither_seed)
        noise = -CENTS_PER_BIN + 2 * CENTS_PER_BIN * np.random.triangular(0.0, 0.5, 1.0, PITCH_BINS)
    finally:
        np.random.set_state(state)
    return CENTS_PER_BIN * np.arange(PITCH_BINS, dtype=np.float64) + CENTS_0 + noise


def _frames(y16000, hop=HOP):
    """Centre-padded 1024-sample frames every `hop` samples, mean-centred and scaled."""
    y = np.asarray(y16000, dtype=np.float32)
    total = 1 + len(y) // hop
    y = np.pad(y, WINDOW // 2)
    idx = np.arange(WINDOW)[None, :] + hop * np.arange(total)[:, None]
    f = y[np.minimum(idx, len(y) - 1)].astype(np.float32)
    f -= f.mean(axis=1, keepdims=True)
    return f / np.maximum(1e-10, f.std(axis=1, ddof=1, keepdims=True))


def _frequency_to_bin(hz, round_fn=np.floor):
    cents = 1200.0 * np.log2(np.asarray(hz, dtype=np.float64) / 10.0)
    return int(round_fn((cents - CENTS_0) / CENTS_PER_BIN))


def f0_track(y16000, device="cpu", dither_seed=0, batch_size=512, hop=HOP):
    """(f0 in Hz, voiced mask), one value per `hop` samples."""
    frames = _frames(y16000, hop)
    net = _model(device)
    out = []
    with torch.no_grad():
        for i in range(0, len(frames), batch_size):
            out.append(net(torch.from_numpy(frames[i:i + batch_size]).to(device)).cpu().numpy())
    probs = np.concatenate(out).astype(np.float64)         # (time, 360), one sigmoid

    lo, hi = _frequency_to_bin(FMIN), _frequency_to_bin(FMAX, np.ceil)
    masked = probs.copy()
    masked[:, :lo] = -np.inf
    masked[:, hi:] = -np.inf

    bins = masked.argmax(axis=1)
    periodicity = np.take_along_axis(masked, bins[:, None], axis=1)[:, 0]

    # weighted average of the dithered bin centres over a +-4 bin window, on a second
    # sigmoid of the already-sigmoid probabilities
    start = np.maximum(0, bins - 4)
    end = np.minimum(PITCH_BINS, bins + 5)
    cols = np.arange(PITCH_BINS)[None, :]
    window = (cols >= start[:, None]) & (cols < end[:, None])
    safe = np.where(window, np.clip(masked, -700.0, 700.0), -700.0)
    p2 = np.where(window, 1.0 / (1.0 + np.exp(-safe)), 0.0)
    cents = (p2 * _bin_cents(dither_seed)[None, :]).sum(axis=1) / p2.sum(axis=1)
    f0 = 10.0 * 2.0 ** (cents / 1200.0)
    return f0.astype(np.float32), periodicity >= CONFIDENCE


def histogram(f0_hz, voiced, tonic_hz, n_bins=N_BINS, smooth=SMOOTH, power=POWER):
    """Voiced frames -> a (n_bins,) octave-folded pitch histogram, summing to 1.

    Unchanged from the reference: it was already plain numpy.
    """
    f0 = np.asarray(f0_hz, dtype=float)[np.asarray(voiced, dtype=bool)]
    cents = 1200.0 * np.log2(np.clip(f0, 1e-6, None) / float(tonic_hz))
    cents = cents[np.isfinite(cents)]
    if cents.size < 5:
        return np.zeros(n_bins)
    idx = np.floor((cents % 1200.0) * (n_bins / 1200.0)).astype(int) % n_bins
    H = np.zeros(n_bins)
    np.add.at(H, idx, 1.0)
    if smooth > 0:
        d = np.arange(n_bins)
        d = np.minimum(d, n_bins - d)
        kern = np.exp(-0.5 * (d / smooth) ** 2)
        H = np.maximum(np.real(np.fft.ifft(np.fft.fft(H) * np.fft.fft(kern / kern.sum()))), 0.0)
    H = H ** power
    total = H.sum()
    return H / total if total > 0 else H
